drivers/usb/bc12/bc12_pi3usb9201.c

```diff
@@ -0,0 +1,392 @@
+				     desired_mode << PI3USB9201_REG_CTRL_1_MODE_SHIFT);
+}
+
+static int pi3usb9201_get_status(const struct device *dev, uint8_t *client, uint8_t *host)
+{
+	const struct pi3usb9201_config *cfg = dev->config;
+	int rv;
+	uint8_t status;
+
+	rv = i2c_reg_read_byte_dt(&cfg->i2c, PI3USB9201_REG_CLIENT_STS, &status);
+	if (rv) {
+		return rv;
+	}
+
+	if (client) {
+		*client = status;
+	}
+
+	rv = i2c_reg_read_byte_dt(&cfg->i2c, PI3USB9201_REG_HOST_STS, &status);
+	if (rv) {
+		return rv;
+	}
+
+	if (host) {
+		*host = status;
+	}
+
+	return 0;
+}
+
+static void pi3usb9201_notify_callback(const struct device *dev, struct bc12_partner_state *state)
+{
+	struct pi3usb9201_data *pi3usb9201_data = dev->data;
+
+	if (pi3usb9201_data->result_callback) {
+		pi3usb9201_data->result_callback(dev, state, pi3usb9201_data->callback_data);
+	}
+}
+
+static void pi3usb9201_update_charging_partner(const struct device *dev,
+					       struct bc12_partner_state *state)
+{
+	struct pi3usb9201_data *pi3usb9201_data = dev->data;
+
+	if (state && state->type == pi3usb9201_data->partner_type) {
+		/* No change to the charging partner */
+		return;
+	}
+
+	/*
+	 * If most recent supplier type is not BC12_TYPE_NONE, then notify
+	 * the callbacks to clear out the BC1.2 state.
+	 */
+	if (pi3usb9201_data->partner_type != BC12_TYPE_NONE) {
+		pi3usb9201_notify_callback(dev, NULL);
+	}
+
+	if (state) {
+		/* Now update the current charger type */
+		pi3usb9201_data->partner_type = state->type;
+		/* If new type != NONE, then notify charge manager */
+		if (state->type != BC12_TYPE_NONE) {
+			pi3usb9201_notify_callback(dev, state);
+		}
+	} else {
+		pi3usb9201_data->partner_type = BC12_TYPE_NONE;
+	}
+}
+
+static int pi3usb9201_client_detect_start(const struct device *dev)
+{
+	int rv;
+
+	/*
+	 * Read both status registers to ensure that all interrupt indications
+	 * are cleared prior to starting bc1.2 detection.
+	 */
+	pi3usb9201_get_status(dev, NULL, NULL);
+
+	/* Put pi3usb9201 into client mode */
+	rv = pi3usb9201_set_mode(dev, PI3USB9201_CLIENT_MODE);
+	if (rv)
+		return rv;
+	/* Have pi3usb9201 start bc1.2 detection */
+	rv = pi3usb9201_bc12_detect_ctrl(dev, true);
+	if (rv)
+		return rv;
+	/* Enable interrupt to wake task when detection completes */
+	return pi3usb9201_interrupt_enable(dev, true);
+}
+
+static void pi3usb9201_client_detect_finish(const struct device *dev, int client_status)
+{
+	struct bc12_partner_state new_chg;
+	int bit_pos;
+
+	/* Set charge voltage to 5V */
+	new_chg.voltage = BC12_CHARGER_VOLTAGE_MV;
+
+	/*
+	 * Find set bit position. Note that this function is only called if a
+	 * bit was set in client_status, so bit_pos won't be negative.
+	 */
+	bit_pos = __builtin_ffs(client_status) - 1;
+
+	new_chg.current = bc12_chg_limits[bit_pos].current_limit;
+	new_chg.type = bc12_chg_limits[bit_pos].partner_type;
+
+	LOG_DBG("client_sts = 0x%x, current = %d mA, type = %d", client_status, new_chg.current,
```

Comment:
It appears zephyr's preferred line length is 100 columns.
https://github.com/zephyrproject-rtos/zephyr/pull/34894

Reply:
> It appears zephyr's preferred line length is 100 columns. #34894

That is a rumor, and clang-format file does not define our coding style, so please fix, **NACK until then.**

Reply:
The [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format) specifically call out using clang-format with the `.clang-format` file for new source files.   It's not reasonable to ask developers to override the `.clang-format` locally.

Reply:
Nope, "The [clang-format tool](https://clang.llvm.org/docs/ClangFormat.html) **can be helpful** to quickly reformat large amounts of new source code". Nothing forces you to use it. If you use it blindly, do not expect your code to be accepted.

And there is no way I will approve code formatted using clang-format that looks ugly like:
```c

	LOG_DBG("client_sts = 0x%x, current = %d mA, type = %d", client_status, new_chg.current_ua,
		new_chg.type);

```
or

```c
		if (!(((msgs[0].flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) && (msgs[0].len == 1) &&
		      ((msgs[1].flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) && (msgs[1].len == 1))) {
			LOG_ERR("Unexpected read msgs");
			return -EIO;
		}

```
Generally, when the line is wrapped at argument or parameter, make it look nice and concise. If the line is too long and you can wrap at argument or parameter, then wrap if it does not worsen readability. You are author of a subsystem, please learn [Zephyr's coding style](https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style) which is actually very simple, do not use stupid tools like clang-format that prevent you from learning and sometimes generates ugly code. Please also do not mark unresolved conversation resolved.

Reply:
> "The [clang-format tool](https://clang.llvm.org/docs/ClangFormat.html) **can be helpful** to quickly reformat large amounts of new source code".

"new source code to our [Coding Style](https://docs.zephyrproject.org/latest/contribute/guidelines.html#id1) standards"

The whole point behind using clang-format is to avoid discussions like this. If style generated by clang-format does not match what is desired in the project, then the project's official clang format configuration should be changed.

Reply:
I'll concede the second example has readability issues, but unrelated to clang format.    I've pushed a new version to improve the readability of the emulator.

The first example is solely a line length debate, and the current .clang-format line limit is 100.